### ISSUE TITLE
Review latest workflow schemas for EOEPCA+

### ISSUE THREAD

**Issue by GarinSmith** [ESA-EarthCODE/open-science-catalog-metadata#233]
EOEPCA+ will look to define a schema for 
1) Reproducible job details  (Workflow Metadata)
2)  Replicable workflow  (Experiment Metadata)

See - System level 
https://github.com/orgs/EOEPCA/projects/4/views/13?sliceBy%5Bvalue%5D=Resource+Discovery&pane=issue&itemId=60227850
 
See - BB level
https://github.com/orgs/EOEPCA/projects/7/views/1?filterQuery=workflow&pane=issue&itemId=69113579

Garin to discuss with @rconway and Angelos whose GitHub id I cannot find yet.
Garin also to confirm that EOEPCA+ Catalogue can ingest and discover OGC API Records. We believe that we will need to write a new Front End in the portal to Find and Access OGC API Records in the same way we do for STAC.

pycsw supports [OGC API - Records - Part 1: Core, version 1.0](https://ogcapi.ogc.org/records) by default.
See https://docs.pycsw.org/en/latest/oarec-support.html
Angelos has confirmed that OSC currently has support to ingest and discover OGC API Records.

Angelos noted that "Open Science Catalog is 2 versions behind, several fixes and new features have been implemented in the last 6 months or so"

Angelos is off from 9 Aug to 9 Sep, but I will meet him tomorrow for his advice on 
EOEPCA+ will look to define a schema for 
1) Reproducible job details  (Workflow Metadata)
2)  Replicable workflow  (Experiment Metadata)

Please also refer to 
https://github.com/orgs/ESA-EarthCODE/projects/5/views/8?pane=issue&itemId=72092886

**Comment by GarinSmith** [comment]
Reviewed with Angelos. Agreed that we should use OGC API Records to 
-   Provide a link to an externally hosted workflow definition
-  This could support numerous workflow types as referenced in the suggested architecture 
    - CWL 
    - openEO
    - Jupyter Notebook    
    - Another etc  
-   Add additionally agreed metadata in a consistent format

E.g. 
openEO  
=====
https://github.com/ESA-APEX/apex_algorithms/blob/main/openeo_udp/worldcereal_inference.json
Link 
"rel": "openeo-process"
"rel": "git"
"rel": "service"
"rel": "license
Additional Metadata
"properties": {
    "created": "2024-05-17T00:00:00Z",
    "updated": "2024-05-17T00:00:00Z",
    "type": "apex_algorithm",
    "title": "ESA worldcereal global maize detector",
    "description": "A maize detection algorithm.",
    "cost_estimate": 0.1,
    "cost_unit": "platform credits per km\u00b2",
etc 


Open Science Catalog
==============
catalog.osc.earthcode.eox.at/collections/metadata:main/items/HCA_L2E_CS_LTA__SIR1SAR_FR_20150331T150158_20150331T150200_D001?f=json
"links"
Addional Metadata
 "properties": {
        "title": "HCA_L2E_CS_LTA__SIR1SAR_FR_20150331T150158_20150331T150200_D001",
        "description": "HYDROCOASTAL Final Product: ........
        "datetime": "2023-02-10T08:45:21.061533Z",
        "start_datetime": "2015-03-31T15:02:32.858513+00:00",
        "end_datetime": "2015-03-31T15:02:34.864426+00:00",
        "created": "2023-02-10T08:45:21.061533+00:00"
    }


Next Steps
1) Review different examples 
2) Angelos to make clearer suggestion on return for standard schema
3) Get feedback from platforms when we know what they are.

**Comment by GarinSmith** [comment]
It is important that this approach also supports the types of workflows identified by @edobrowolska in 
https://github.com/ESA-EarthCODE/portal/issues/17

Ewelina, identified a number of scripts that can be considered workflows. E.g 
https://github.com/diarmuidcorr/Lake-Channel-Identifier/blob/v1.0/Landsat-8%20SGL%20and%20Channel%20Classifier  (Python script)
https://github.com/GEUS-SICE/SICE/blob/master/S3_wrapper.sh    (Python script)

These scripts might be regarded as unstructured workflows, perhaps like a Jupyter Notebook.  It may be that at some point they might be converted to a more formal workflow like for instance a CWL file (OGC API Processes). However there is no reason why these unstructured scripts cannot be used and supported by EarthCODE using the above approach. 

E.g.

    {
      "rel": "git",
      "type": "application/json",
      "title": "Git source repository",
      "href": "https://github.com/diarmuidcorr/Lake-Channel-Identifier/blob/v1.0/Landsat-8%20SGL%20and%20Channel%20Classifier"
    },

 or

    {
      "rel": "git",
      "type": "application/json",
      "title": "Git source repository",
      "href": "https://github.com/GEUS-SICE/SICE/blob/master/S3_wrapper.sh"
    },

The above syntax may not be 100% correct, but hopefully, it demonstrates what is possible with OGC API Records. 

It may be that 
i) Angelos can recommend an elaborated standard approach.
ii)  One/all of the chosen EarthCODE contractors  can suggest a suitable schema that already works with the existing platforms .
iii) A combination of the above works.

**Comment by GarinSmith** [comment]
We need to confirm the schema that will be used for validation of OGC API  Records
See  https://github.com/opengeospatial/ogcapi-records/tree/master/core/openapi/schemas
E.g.
https://github.com/opengeospatial/ogcapi-records/blob/master/core/openapi/schemas/recordJSON.yaml

Hopefully we can test some of the above  examples using the correct schema.

**Comment by GarinSmith** [comment]
I just reviewed this approach with EOEPCA+ and will link then to this user story to help clarify our requirements. 

I would like EOEPCA+ guidance on how to validate schema compliance?  This seems quite complicated. 
E.g. see  
https://json-schema.org/implementations#validators-web-(online) 
or
https://json-schema.org/implementations#command-line

I have looked at command lined solutions like  Polyglottal JSON Schema Validator
E.g. 
pajv validate -s recordGeoJSON.yaml -d record.json -r recordCommonProperties.yaml -r time.yaml -r linkBase.yaml -r linkTemplate.yaml
but this does not yet seem to work.

**Comment by GarinSmith** [comment]
I also got some useful strategic possibilities from EOEPCA 

E.g.  
from  Jonas Sølvsteen
( potential platform integration with UK EO Data Hub 
https://github.com/os-climate/hazard/blob/main/hazard_workflow.cwl
(this is meant to run among others on the UK EO Data Hub https://eodatahub.org.uk/)
 https://d1fzab3z0mlfhy.cloudfront.net/
 https://radiantearth.github.io/stac-browser/#/external/pgstac.demo.cloudferro.com/
 
from  Gérald FENOY
https://ospd-02.geolabs.fr/examples/cwls/algae-usecase-workflow-copernicus.cwl
https://ospd-02.geolabs.fr/examples/app-package.cwl

**Comment by GarinSmith** [comment]
Angelos is now back and will ask Peter Vretanos from the OGC API Records SWG about the tool he used to validate all the examples in the specification.

**Comment by kalxas** [comment]
https://github.com/EOEPCA/resource-discovery/issues/56#issuecomment-2360542313

**Comment by GarinSmith** [comment]
The link above provided by @kalxas is very helpful and provides a very good starting point to move forwards for EarthCODE.

**Comment by GarinSmith** [comment]
I have now successfully validated the OpenEO example above using a command line tool. 
The key parts of this investigation are stored in the updated Architecture document. 

In summary 
1) We have a tool to validate OGC API Records (https://github.com/wmo-im/wcmp2/tree/main/schemas)
2) I can successfully validate the OpenEO example above although I had to fix 3 minor problems
time must be set as null - "time": null
geometry must be set as null - "geometry": null 
"organization" is mandatory and was missing in one instance.
3) We could use the OpenEO for other workflows. See Other Examples below
4) All these examples also validate using the tool above. 
5) This is a good starting point, but we need to agree with Angelos what is the best way to 
i) Know what a workflow type is somehow.
E.g. openeo, ogc api processes, JNB, free format Python and so on.
ii) Know if a processes is a Workflow (generic) , Experiment (specific) or Dashboard (GUI to Workflow).
Apparently the examples below may not be best practise although they do validate and they also conform to the current OpenEO approach, so in a way they are perhaps good reference/starting point. 
I have asked Angelos for guidance.

**Other Examples**

OpenEO 

links": [
  {
    "rel": "openeo-process",
    "type": "application/json",
    "title": "openEO Process Title",
    "href": "https://raw.githubusercontent.com/ESA-APEx/apex_algorithms/max_ndvi_composite/openeo_udp/examples/max_ndvi_composite/max_ndvi_composite.json"
  }

OGC API Processes

links": [
  {
    "rel": "ogcapi-process",
    "type": "application/json",
    "title": "OGC API Process Title",
    "href": "https://owncloud.spaceapplications.com/owncloud/index.php/s/iCk60Kmry77o2l6/download"
  }

Python  Processes  

links": [
  {
    "rel": "python-process",
    "type": "application/json",
    "title": "Python Process Title",
    "href": "https://github.com/GEUS-SICE/SICE/blob/master/S3_wrapper.sh"
  }

Jupyter Notebook Processes

links": [
  {
    "rel": "jupyter-notebook",
    "type": "application/json",
    "title": "Python Process Title",
    "href": "https://github.com/...../..../file.ipynb"
  }

**Comment by kalxas** [comment]
@GarinSmith please check the new EOEPCA metadata schema here:
https://github.com/EOEPCA/metadata-profile/tree/master/schemas

**Comment by GarinSmith** [comment]
@kalxas i just looked, 
Is it possible to give us the schemas/resource-bundled.json  ?
to avoid having to install [Stoplight Studio](https://stoplight.io/studio)  etc
and we can then re-test our existing ogc api records JSON examples?
See other thread regarding the suggestion of focusing on potential real world examples to help make progress.
Please see    https://github.com/EOEPCA/resource-discovery/issues/56

**Comment by GarinSmith** [comment]
Thanks, I have used this new schema, but I am having trouble converting the EOPECA example to one that works
please see  https://github.com/EOEPCA/resource-discovery/issues/56

1) Could you please convert the EarthCODE example to one that validates? So we then have a good working reference.
2) Could you explain how we will know what type of workflow we are dealing with (openeo-process in this case)?
E.g.
openeo-process
ogcapi-process
JNB
etc
This may be clear when 1) is answered.

**Comment by GarinSmith** [comment]
It would also help to have a reference guide such as
https://schemas.wmo.int/wcmp/2.0.0/standard/wcmp-2.0.0.pdf
although this may be less necessary  when we have a working reference (see above).

**Comment by Schpidi** [comment]
If I understand this correctly it would be great to prepare a PR towards https://github.com/ESA-EarthCODE/open-science-catalog-validation to add the schema and any other validation artifacts extending the validation command `open-science-catalog-validation ./{eo-missions,products,projects,themes,variables}`.

The consecutive task would be to add a first example workflow to https://github.com/ESA-EarthCODE/open-science-catalog-metadata and extend the validation action with the extended command.

**Comment by kalxas** [comment]
@GarinSmith 
Let's see what we did for Processes, ADES, CSW in EOEPCA v1.x:
We used the type "service" and added keywords in the record to specify that it is of a specific flavor.
https://resource-catalogue.develop.eoepca.org/collections/metadata:main/items/https---demo-pygeoapi-io-stable-processes-
For CWL records, we used type=application and provided CWL in keywords.
We did not use a custom link relation in all those cases.

I propose we follow the same pattern, using record.type to specify workflows.

There are 2 options:
1. Use type="workflow" for all flavors of workflows and specify the flavor in a keyword (that makes it queryable in the catalog)
2. Use the flavor as a dedicated type (e.g. type="openeo") and add the keyword "workflow"

**Comment by silvester-pari** [comment]
@kalxas there are some issues with the schema https://github.com/EOEPCA/metadata-profile/blob/master/schemas/resource.json, like the given URL and the $id in the schema are not consistent (e.g. yaml file extension instead of .json) and some other things that the Node-based validator (AJV) we are using is complaining about (seems to be stricter than the Python one check-jsonschema). What's the best way to iterate on this schema to make it work for OSC?

**Comment by m-mohr** [comment]
Yeah, the schema should be hosted in a way that the $id matches the actual URL and ideally it would also return a JSON media type instead of text/plain. Maybe host it through GitHub Pages?

**Comment by kalxas** [comment]
The schema was generated from the OGC API - Records yaml schema using the tools described here: https://github.com/EOEPCA/metadata-profile/tree/master/schemas
The upstream schema is here:
https://github.com/opengeospatial/ogcapi-records/blob/master/core/openapi/schemas/recordGeoJSON.yaml

**Comment by m-mohr** [comment]
Yeah OpenAPI (3.0) and JSON Schema can't be converted 1:1, it seems the tool doesn't do the conversion very strictly. Shall we propose the changes from my Gist to the upstream schema then?

**Comment by kalxas** [comment]
thanks @m-mohr , opened an upstream pull request

**Comment by silvester-pari** [comment]
Thanks @kalxas, there was a second issue:
> AJV also complains about the keyword "example" which should be "examples" (and an array). 

Is this also something that can be fixed upstream?

**Comment by kalxas** [comment]
I will check with the SWG

**Comment by silvester-pari** [comment]
@kalxas any updates yet?

**Comment by m-mohr** [comment]
Shouldn't the example provided above (i.e. https://schemas.wmo.int/wcmp/2.0.0/examples/de-dwd.surface-weather-observations-realtime.json )  include a conformance class for (static) OGC API - Records?

**Comment by kalxas** [comment]
no further updates yet

**Comment by kalxas** [comment]
I think OpenAPI allows for both example (string/object/array) and examples (array)

**Comment by m-mohr** [comment]
That's correct, but then it needs to be converted to examples for JSON Schema. So why not just use examples also in OpenAPI if that's the common denominator? Makes it easier to switch between JSON Schema and OpenAPI.

**Comment by kalxas** [comment]
https://github.com/opengeospatial/ogcapi-records/issues/396

**Comment by kalxas** [comment]
@m-mohr regarding conformance classes:
If the conformance class is based on a requirements class that has a dependency, why should we additionally specify OARec as another conformance class?

**Comment by m-mohr** [comment]
@kalax How can I resolve in a client which conformance classes apply including all depencies? My client can't know all conformance classes of the whole world and it's dependencies, but should be able validate whether something is a valid Record. It doesn't know whether it's Records though because the conformace class is not listed. So it would reject the file as invalid Records. So I think the dependencies need to be listed. I would rather ask: Why would you not list it?

**Comment by kalxas** [comment]
Just out of the OGC API Records meeting, we have covered all the pending issues. Changes to be applied shortly as pull requests.

**Comment by m-mohr** [comment]
Can you remind me what the solution for the example vs examples issue was? @kalxas

**Comment by kalxas** [comment]
Yes, there is a plan to migrate OGC APIs to OpenAPI 3.1

**Comment by kalxas** [comment]
https://github.com/EOEPCA/metadata-profile/pull/2

**Comment by kalxas** [comment]
The example(s) have been removed from the schema (also upstream)

**Comment by Schpidi** [comment]
Updated PR for records validation in https://github.com/ESA-EarthCODE/open-science-catalog-validation/pull/12
Tested against a draft example included in https://github.com/ESA-EarthCODE/open-science-catalog-metadata-testing/tree/demo-workflow

**Comment by m-mohr** [comment]
Records validation now in: https://github.com/ESA-EarthCODE/open-science-catalog-validation/pull/16
Example metadata in: https://github.com/ESA-EarthCODE/open-science-catalog-metadata-testing/pull/69
